### ISSUE TITLE
symlink busybox /busybox exists

### ISSUE THREAD

**Issue by lambertjosh** [GoogleContainerTools/kaniko#232]
As a simple test, I was trying to use kaniko to build itself. I ran into an issue however that it failed unpacking the images, because `/busybox/` already existed.

Note I am using a slightly modified `kaniko:latest` image to get it to work with GitLab CI, but I don't think that is a direct cause of this issue here. It seems rather that this directory is not whitelisted, but should be?

```
time="2018-07-10T15:02:52Z" level=info msg="Unpacking layer: 6"
time="2018-07-10T15:02:52Z" level=info msg="Not adding /kaniko because it is whitelisted"
time="2018-07-10T15:02:52Z" level=info msg="Not adding /kaniko/.config because it is whitelisted"
time="2018-07-10T15:02:52Z" level=info msg="Not adding /kaniko/.config/gcloud because it is whitelisted"
time="2018-07-10T15:02:52Z" level=info msg="Not adding /kaniko/.config/gcloud/docker_credential_gcr_config.json because it is whitelisted"
time="2018-07-10T15:02:52Z" level=info msg="Unpacking layer: 5"
time="2018-07-10T15:02:52Z" level=info msg="Not adding /kaniko because it is whitelisted"
time="2018-07-10T15:02:52Z" level=info msg="Not adding /kaniko/.docker because it is whitelisted"
time="2018-07-10T15:02:52Z" level=info msg="Not adding /kaniko/.docker/config.json because it is whitelisted"
time="2018-07-10T15:02:52Z" level=info msg="Unpacking layer: 4"
time="2018-07-10T15:02:52Z" level=info msg="Not adding /kaniko because it is whitelisted"
time="2018-07-10T15:02:52Z" level=info msg="Not adding /kaniko/ssl because it is whitelisted"
time="2018-07-10T15:02:52Z" level=info msg="Not adding /kaniko/ssl/certs because it is whitelisted"
time="2018-07-10T15:02:52Z" level=info msg="Not adding /kaniko/ssl/certs/ca-certificates.crt because it is whitelisted"
time="2018-07-10T15:02:52Z" level=info msg="Unpacking layer: 3"
time="2018-07-10T15:02:52Z" level=error msg="symlink busybox /busybox/[: file exists"
```

**Comment by priyawadhwa** [comment]
The reason you're seeing this error is because `/busybox` already exists, so when kaniko tries to symlink `/busybox` from the base image there's a symlink error. This is why we generally recommend using the standard executor image, since adding things like a shell might prevent kaniko from building certain images (although I understand you need one for it to work with GitLab!).

If we move the `/busybox` directory to `/kaniko/busybox` in the debug image it would be whitelisted and this might work (although I haven't tried yet!)

**Comment by pmcq** [comment]
Hi, 
I ran into a similar issue building a simple image based on `ubuntu:16.04`.  The simplest reproduction I could find is a Dockerfile just containing `FROM ubuntu:16.04`.

I end up with the following error:
```
DEBU[0035] symlink from hostname to /bin/dnsdomainname  
ERRO[0035] symlink hostname /bin/dnsdomainname: file exists 
```
However, after I poked around some more (added some debug code) I realized that when creating symlinks it's possible the file that is being linked to does not yet exist in the layer, and that the error message says `file exists` but it does not.
https://github.com/GoogleContainerTools/kaniko/blob/master/pkg/util/fs_util.go#L221-L232 

Debug code I added to L227:
```
if _, err := os.Stat(hdr.Linkname); os.IsNotExist(err) {
     logrus.Debugf("link does not yet exist! %s", hdr.Linkname)
}
```
```
DEBU[0034] symlink from hostname to /bin/dnsdomainname  
DEBU[0034] link does not yet exist! hostname            
ERRO[0034] symlink hostname /bin/dnsdomainname: file exists 
```

I believe a possible fix would be to add all symlink/hardlink types to an array in this loop https://github.com/GoogleContainerTools/kaniko/blob/master/pkg/util/fs_util.go#L67 and only try extracting the files after all regular files/directories have been extracted.  Does that seem reasonable?
